usr.sbin/bhyve/migration.c

```diff
@@ -573,115 +534,86 @@ migrate_send_memory(struct vmctx *ctx, int socket)
 	return (0);
 }
 
+/**
+ * The source host saves the state for the kernel structure that will be
+ * migrated and sends to the destination host a message that contains
+ * the type of data to be sent (MESSAGE_TYPE_KERN), the size of the structure
+ * to be received and the index that represents the kernel structure in order to
+ * be identified by the destination host. Then, the source host transfer the
+ * state of the kernel structure over the network and the destination host
+ * restores it.
+ */
 static inline int
-migrate_send_kern_struct(struct vmctx *ctx, int socket,
-			char *buffer,
-			enum snapshot_req struct_req)
+migrate_kern_struct(struct vmctx *ctx, int socket, char *buffer,
+		    enum snapshot_req struct_req, enum migration_transfer_req req)
 {
 	int rc;
 	size_t data_size;
 	struct migration_message_type msg;
 	struct vm_snapshot_meta *meta;
 
-	memset(&msg, 0, sizeof(msg));
-	msg.type = MESSAGE_TYPE_KERN;
-
-	meta = &(struct vm_snapshot_meta) {
-		.ctx = ctx,
+	if ((req != MIGRATION_SEND_REQ) && (req != MIGRATION_RECV_REQ)) {
+		fprintf(stderr, "%s: Unknown request\r\n", __func__);
+		return (-1);
+	}
 
-		.dev_req = struct_req,
+	memset(&msg, 0, sizeof(msg));
+	if (req == MIGRATION_SEND_REQ) {
+		msg.type = MESSAGE_TYPE_KERN;
 
-		.buffer.buf_start = buffer,
-		.buffer.buf_size = SNAPSHOT_BUFFER_SIZE,
+		meta = ALLOCA_VM_SNAPSHOT_META(ctx, NULL, struct_req, buffer, SNAPSHOT_BUFFER_SIZE, VM_SNAPSHOT_SAVE);
+		memset(meta->buffer.buf_start, 0, meta->buffer.buf_size);
+		meta->buffer.buf = meta->buffer.buf_start;
+		meta->buffer.buf_rem = meta->buffer.buf_size;
```

Comment:
The` ALLOCA_VM_SNAPSHOT ` "returns" only the allocation of the struct `vm_snapshot_meta *` variable. (e.g. `meta = ALLOCA_VM_SNAPSHOT();`)

Since the macro can be moved and, maybe, used by the snapshot feature, the `meta->buffer.buf` and `meta->buffer.buf_rem` may be initialized in other manner.